### PR TITLE
Audio pertubation for switchboard experiments

### PR DESCRIPTION
**Pull request by Max-Ryujin**
`perturbation-py` is nearly identical to  `recipe/crnn/helpers/vieting/perturbation.py`

`baseline_args.py` added a parameter to add the Code to the config.

### CODE REVIEW

File: users/vieting/experiments/switchboard/hybrid/feat/experiments.py [hunk 1]
```diff
@@ -299,9 +307,40 @@ def run_specaug_scf():
...
+        num_epochs=260,
+    )
+    nn_steps = RasrSteps()
+    nn_steps.add_step("nn", nn_args)
+
+    hybrid_nn_system = get_hybrid_nn_system(context_window=249, audio_opts=audio_opts) 
+    hybrid_nn_system.run(nn_steps)
+    for train_job in hybrid_nn_system.jobs["switchboard.train_switchboard.cv"].values():
+        # noinspection PyUnresolvedReferences
+        train_job.rqmt.update({"gpu_mem": 24, "mem": 10})
```
Comment (vieting) [inline users/vieting/experiments/switchboard/hybrid/feat/experiments.py:339]: Can we also run this on 11GB GPUs, possibly with gradient accumulation of 2?

File: users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+class WaveformPerturbation:
+  """
+  Helper class to perform perturbation techniques on audio waveforms.
+  """
+  def __init__(self, speed=None, tempo=None, sox_effects=None, codecs=None, preemphasis=None):
+    """
+    :param speed:
+    """
+    from functools import partial
```
Comment (vieting) [inline users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py:25]: Below, `functools.partial` is used.

File: users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+      self._perturbations.append(functools.partial(self.apply_codecs, codecs=codecs))
+
+  def run(self, audio, sample_rate, random_state):
+    # input_shape = audio.shape
+    audio = torch.from_numpy(audio).unsqueeze(0).to(torch.float32)
+    for perturbation in self._perturbations:
+      audio = perturbation(audio, sample_rate, random_state)
+    audio = audio.numpy().squeeze()
+    # output_shape = audio.shape
+    # print("change shape from {} to {}".format(input_shape, output_shape))
```
Comment (vieting) [inline users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py:41]: This can be removed.
```suggestion
```

File: users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    self._speed = PerturbationFactor(**speed) if speed else None
+    self._tempo = PerturbationFactor(**tempo) if tempo else None
+    self._perturbations = [functools.partial(self.sox, sox_effects=sox_effects)]
+    if preemphasis:
+      self._perturbations.append(functools.partial(self.preemphasis, factor=PerturbationFactor(**preemphasis)))
+    if codecs:
+      self._perturbations.append(functools.partial(self.apply_codecs, codecs=codecs))
+
+  def run(self, audio, sample_rate, random_state):
+    # input_shape = audio.shape
```
Comment (vieting) [inline users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py:35]: This can be removed.
```suggestion
```

File: users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+  @staticmethod
+  def preemphasis(audio, sample_rate, random_state, factor):
+    if random_state.random() < factor.prob:
+      preemphasis_coefficient = random_state.random() * (factor.max - factor.min) + factor.min
+      # audio[i,j] -= preemphasis_coefficient * audio[i, max(0, j-1)] for all i,j
+      offset_audio = torch.nn.functional.pad(audio.unsqueeze(0), (1, 0), mode="replicate").squeeze(
+          0
+      )  # size (m, window_size + 1)
+      audio = audio - preemphasis_coefficient * offset_audio[:, :-1]
```
Comment (vieting) [inline users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py:70]: I'm not sure how exactly this materialized, but it seems like torchaudio has a nice functionality for preemphasis which we can use instead, see here in the [docs](https://pytorch.org/audio/stable/generated/torchaudio.functional.preemphasis.html).
```suggestion
      audio = torchaudio.functional.preemphasis(audio, coeff=preemphasis_coefficient)
```

File: users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    if self._speed is not None:
+      if random_state.random() < self._speed.prob:
+        factor = random_state.random() * (self._speed.max - self._speed.min) + self._speed.min
+        sox_effects.append(["speed", str(factor)])
+        sox_effects.append(["rate", str(sample_rate)])
+        speed = True
+    if self._tempo is not None:
+      if random_state.random() < self._tempo.prob and not speed:
+        factor = random_state.random() * (self._tempo.max - self._tempo.min) + self._tempo.min
+        sox_effects.append(["tempo", str(factor)])
```
Comment (vieting) [inline users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py:58]: Right now this seems a bit strange to me. This function might be called multiple times and each time, the `sox_effects` will be appended, right?

**Review by vieting — CHANGES_REQUESTED** [review]

File: users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py (hunk 3 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py]: Done

File: users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py (hunk 5 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py]: Changed

File: users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py (hunk 4 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py]: Done

File: users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py (hunk 2 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py]: Chnanged import to import functools

File: users/vieting/experiments/switchboard/hybrid/feat/experiments.py (hunk 1 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/hybrid/feat/experiments.py]: requires testing. I changed it to 11GB for now

File: users/vieting/experiments/switchboard/hybrid/feat/experiments.py [hunk 7]
```diff
@@ -299,9 +307,40 @@ def run_specaug_scf():
         # noinspection PyUnresolvedReferences
         train_job.rqmt.update({"gpu_mem": 24, "mem": 10})
 
+def run_audio_perturbation_scf():
+    gs.ALIAS_AND_OUTPUT_SUBDIR = "experiments/switchboard/hybrid/feat/"
+    audio_opts = {"features": "raw", "peak_normalization": True, "pre_process": CodeWrapper("audio_perturb_runner.run")}
+    nn_args = get_nn_args_baseline(
+        nn_base_args={
+            "scf": dict(returnn_args=dict(
+                batch_size=14000,
```
[the remark this answers is not in the document]
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/hybrid/feat/experiments.py:311]: Not yet.

File: users/vieting/experiments/switchboard/hybrid/feat/experiments.py [hunk 8]
```diff
@@ -265,13 +266,19 @@ def run_specaug_scf():
 
     nn_args = get_nn_args_baseline(
         nn_base_args={
-            "scf": dict(returnn_args=dict(batch_size=7000, specaug_mask_sorting=True, extra_args=dict(accum_grad_multiple_step=2)),
+            "scf": dict(returnn_args=dict(
+                batch_size=3500,
+                specaug_mask_sorting=True,
+                specaug_after_first_layer=True,
```
[the remark this answers is not in the document]
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/hybrid/feat/experiments.py]: Yes, the main version is not correct. specaug_mask_sorting requires now specaug_after_first_layer. Else you get an assertion error.

File: users/vieting/experiments/switchboard/hybrid/feat/experiments.py [hunk 9]
```diff
@@ -265,13 +266,19 @@ def run_specaug_scf():
 
     nn_args = get_nn_args_baseline(
         nn_base_args={
-            "scf": dict(returnn_args=dict(batch_size=7000, specaug_mask_sorting=True, extra_args=dict(accum_grad_multiple_step=2)),
+            "scf": dict(returnn_args=dict(
+                batch_size=3500,
```
[the remark this answers is not in the document]
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/hybrid/feat/experiments.py]: The values are not correct. 3500 is required for the training to work. Would change it together with the fix below in another PR.

File: users/vieting/experiments/switchboard/hybrid/feat/experiments.py [hunk 10]
```diff
@@ -272,6 +273,7 @@ def run_specaug_scf():
                 returnn_args=dict(
                     batch_size=3500, 
                     specaug_mask_sorting=True,
+                    specaug_after_first_layer=True,
```
Comment (vieting) [inline users/vieting/experiments/switchboard/hybrid/feat/experiments.py:276]: As discussed in other threads, please move this to a separate PR.
```suggestion
```

File: users/vieting/experiments/switchboard/hybrid/feat/experiments.py [hunk 11]
```diff
@@ -281,6 +283,7 @@ def run_specaug_scf():
                 returnn_args=dict(
                     batch_size=3500,
                     specaug_mask_sorting=True,
+                    specaug_after_first_layer=True,
```
Comment (vieting) [inline users/vieting/experiments/switchboard/hybrid/feat/experiments.py:286]: As discussed in other threads, please move this to a separate PR.
```suggestion
```

File: users/vieting/experiments/switchboard/hybrid/feat/experiments.py [hunk 12]
```diff
@@ -299,9 +302,40 @@ def run_specaug_scf():
...
+        num_epochs=260,
+    )
+    nn_steps = RasrSteps()
+    nn_steps.add_step("nn", nn_args)
+
+    hybrid_nn_system = get_hybrid_nn_system(context_window=249, audio_opts=audio_opts) 
+    hybrid_nn_system.run(nn_steps)
+    for train_job in hybrid_nn_system.jobs["switchboard.train_switchboard.cv"].values():
+        # noinspection PyUnresolvedReferences
+        train_job.rqmt.update({"gpu_mem": 11, "mem": 10})
```
Comment (vieting) [inline users/vieting/experiments/switchboard/hybrid/feat/experiments.py:334]: The 11 GB GPU memory are requested by default. If you want to use these GPUs, this is not needed.

File: users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+    audio, _ = torchaudio.sox_effects.apply_effects_tensor(audio, sample_rate, sox_effects)
+    return audio
+
+  @staticmethod
+  def preemphasis(audio, sample_rate, random_state, factor):
+    import torch
```
Comment (vieting) [inline users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py:59]: I guess it should be `torchaudio` here? In general, I think it would be a bit nicer not to have the imports in each function but once at the beginning. You could achieve this by adding a string with the imports as the first entry in the list returned by `get_classes_perturbation()` (which might then be renamed to better reflect what it does).

File: users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+
+
+class WaveformPerturbation:
+  """
+  Helper class to perform perturbation techniques on audio waveforms.
+  """
+  def __init__(self, speed=None, tempo=None, sox_effects=None, codecs=None, preemphasis=None):
+    """
+    :param speed:
+    """
```
Comment (vieting) [inline users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py:18]: This docstring is incomplete. It would be nice to add some short description and examples for each argument.

File: users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+    self.prob = prob
+    self.min = minimum
+    self.max = maximum
+
+
+class WaveformPerturbation:
+  """
+  Helper class to perform perturbation techniques on audio waveforms.
+  """
+  def __init__(self, speed=None, tempo=None, sox_effects=None, codecs=None, preemphasis=None):
```
Comment (vieting) [inline users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py:15]: Please add type annotations.

**Review by vieting — CHANGES_REQUESTED** [review]

File: users/vieting/experiments/switchboard/hybrid/feat/experiments.py (hunk 10 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/hybrid/feat/experiments.py]: Done

File: users/vieting/experiments/switchboard/hybrid/feat/experiments.py (hunk 11 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/hybrid/feat/experiments.py]: Done

File: users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py (hunk 13 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py]: Changed it so the imports are added separately

File: users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py (hunk 14 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py]: Done

File: users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py (hunk 15 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py]: Done

File: users/vieting/experiments/switchboard/hybrid/feat/experiments.py (hunk 1 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/hybrid/feat/experiments.py]: Not relevant anymore, since audio_perturbation moed to CTC

File: users/vieting/experiments/switchboard/hybrid/feat/experiments.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/hybrid/feat/experiments.py:309]: Not relevant anymore, since audio_perturbation moed to CTC

File: users/vieting/experiments/switchboard/hybrid/feat/experiments.py (hunk 12 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/hybrid/feat/experiments.py]: Removed

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py [hunk 16]
```diff
@@ -22,6 +22,12 @@
 from .fullsum_ctc_raw_samples import make_conformer_fullsum_ctc_model
 from .network_helpers.learning_rates import oclr_default_schedule
 
+from .network_helpers.perturbation import (
+    PerturbationFactor,
+    WaveformPerturbation,
+    get_code_for_perturbation,
+)
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:29]:
```suggestion
from .network_helpers.perturbation import get_code_for_perturbation
```
Seems like this is already sufficient, right?

File: users/vieting/experiments/switchboard/ctc/feat/conformer_baseline.py [hunk 17]
```diff
@@ -154,6 +155,25 @@ def run_test_mel():
...
+                        "audio_perturb_runner": CodeWrapper("WaveformPerturbation(**audio_perturb_args)")
+                    },
+                    **returnn_args
+                },
+                feature_args=feature_args,
+                lr_args={
+                    "peak_lr": 4e-4, "start_lr": 1.325e-05, "end_lr": 1e-5,
+                    "increase_epochs": 119, "peak_epochs": 2, "decrease_epochs": 119, "final_epochs": 0,
+                },
+                report_args={"architecture": "conf-wei", "lr": "wei_peak_4e-4"},
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/conformer_baseline.py:175]:
```suggestion
                report_args={"architecture": "conf-wei", "lr": "wei_peak_4e-4", "speed": "0.6_0.88_1.12", "tempo": "0.6_0.83_1.17"},
```
Something like this will add information about the perturbation to the csv report that is created. Please check if the line is too long.

File: users/vieting/experiments/switchboard/ctc/feat/conformer_baseline.py [hunk 18]
```diff
@@ -154,6 +155,25 @@ def run_test_mel():
                 },
                 report_args={"architecture": "conf-wei", "lr": "wei_peak_4e-4", "specaug": "wei_adapt_80dim"},
             ),
+            "lgm80_conf-audio_perturbation": dict(
+                returnn_args={
+                    "conformer_type": "wei",
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/conformer_baseline.py:160]:
```suggestion
                    "conformer_type": "wei",
                    "specaug_old": {},
```
This was better so far.

File: users/vieting/experiments/switchboard/ctc/feat/conformer_baseline.py [hunk 19]
```diff
@@ -154,6 +155,25 @@ def run_test_mel():
                 },
                 report_args={"architecture": "conf-wei", "lr": "wei_peak_4e-4", "specaug": "wei_adapt_80dim"},
             ),
+            "lgm80_conf-audio_perturbation": dict(
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/conformer_baseline.py:158]:
```suggestion
            "lgm80_conf-wei-oldspecaug-audio_perturbation": dict(
```

File: users/vieting/experiments/switchboard/ctc/feat/conformer_baseline.py [hunk 20]
```diff
@@ -301,3 +321,5 @@ def run_test_mel():
         os.path.join(gs.ALIAS_AND_OUTPUT_SUBDIR, "report.csv"),
         values=report.get_values(),
         template=report.get_template())
+
+
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/conformer_baseline.py:325]: What are these for?

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+    def apply_codecs(audio, sample_rate, random_state, codecs):
+        for codec in codecs:
+            prob = codec.pop("prob", 1.0)
+            if random_state.random() < prob:
+                audio = torchaudio.functional.apply_codec(audio, sample_rate, **codec)
+        return audio
+
+
+def get_code_for_perturbation():
+    classes = ["import torch", "import numpy as np", "import torchaudio", "import functools", "import random", "from typing import List, Dict"]
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:124]: Isn't this line too long?

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+        - 'maximum' (float): The maximum factor by which the audio speed will be increased.
+        Example: {"prob": 0.6, "minimum": 0.88, "maximum": 1.12}
+
+    :param tempo: A dictionary specifying the parameters for tempo perturbation.
+        - 'prob' (float): The probability of applying tempo perturbation. 
+        - 'minimum' (float): The minimum factor by which the audio tempo will be decreased.
+        - 'maximum' (float): The maximum factor by which the audio tempo will be increased.
+        Example: {"prob": 0.6, "minimum": 0.83, "maximum": 1.17}
+
+    :param sox_effects: A list of dictionaries, each dictionary representing a SoX effect.
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:36]: I think it is a list of lists, isn't it?

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+        :param speed: A dictionary containing parameters for the speed perturbation.
+            Expected keys:
+            - factor (float): The factor by which the audio speed will be perturbed.
+        :param tempo: A dictionary containing parameters for the tempo perturbation.
+            Expected keys:
+            - factor (float): The factor by which the audio tempo will be perturbed.
+        :param sox_effects: A list of dictionaries, each containing parameters for a specific SoX effect.
+        :param codecs: A list of dictionaries, each containing parameters for applying codecs to the audio.
+        :param preemphasis: A dictionary containing parameters for the preemphasis filter.
+        """
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:73]: This docstring is quite redundant with the one above. I'd suggest to move the documentation of the parameters here and just have it once.

File: users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py [hunk 24]
```diff
@@ -0,0 +1,128 @@
+import functools
```
Comment (vieting) [inline users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py:1]: Please remove it from here for now.

**Review by vieting — CHANGES_REQUESTED** [review]

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py (hunk 22 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:36]: I don't think so.

File: users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py (hunk 24 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py]: Removed and commented the relevant parts in experiments.py out.

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py (hunk 23 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py]: Done

File: users/vieting/experiments/switchboard/ctc/feat/conformer_baseline.py (hunk 20 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/conformer_baseline.py]: Removed

File: users/vieting/experiments/switchboard/hybrid/feat/baseline_args.py [hunk 25]
```diff
@@ -26,7 +26,7 @@
     specaug_layer_sorted,
     get_funcs_sorted,
 )
-
+
```
Comment (vieting) [inline users/vieting/experiments/switchboard/hybrid/feat/baseline_args.py:29]: If you push code, please also check the diff or go through the PR here on GitHub. Then you see unneeded changes like this one.

File: users/vieting/experiments/switchboard/hybrid/feat/experiments.py [hunk 26]
```diff
@@ -299,9 +300,37 @@ def run_specaug_scf():
...
+            #),
+        #},
+        #prefix="conformer_bs14k_audio_perturbation_",
+        #num_epochs=260,
+    #)
+    #nn_steps = RasrSteps()
+    #nn_steps.add_step("nn", nn_args)
+
+    #hybrid_nn_system = get_hybrid_nn_system(context_window=249, audio_opts=audio_opts) 
+    #hybrid_nn_system.run(nn_steps)
```
Comment (vieting) [inline users/vieting/experiments/switchboard/hybrid/feat/experiments.py:329]: We shouldn't add commented code here to main.
```suggestion
```

File: users/vieting/experiments/switchboard/hybrid/feat/experiments.py [hunk 27]
```diff
@@ -3,6 +3,7 @@
 from sisyphus import gs, tk
 
 from i6_core.features.common import samples_flow
+from i6_core.returnn.config import CodeWrapper
```
Comment (vieting) [inline users/vieting/experiments/switchboard/hybrid/feat/experiments.py:6]:
```suggestion
```

File: users/vieting/experiments/switchboard/hybrid/feat/experiments.py [hunk 28]
```diff
@@ -299,9 +300,37 @@ def run_specaug_scf():
...
+
+    #hybrid_nn_system = get_hybrid_nn_system(context_window=249, audio_opts=audio_opts) 
+    #hybrid_nn_system.run(nn_steps)
 
 def run_all():
     run_baseline_gt()
     run_baseline_mel()
     run_baseline_scf()
     run_specaug_scf()
+    #run_audio_perturbation_scf()
```
Comment (vieting) [inline users/vieting/experiments/switchboard/hybrid/feat/experiments.py:336]:
```suggestion
```

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py (hunk 22 above)
Comment (vieting) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py]: I do think so :smile:  See the [docs](https://pytorch.org/audio/main/generated/torchaudio.sox_effects.apply_effects_tensor.html).

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py [hunk 29]
```diff
@@ -0,0 +1,123 @@
...
+    SoX effects, codec application, and pre-emphasis filtering. 
+    The parameters `speed`, `tempo`, `codecs`, and `preemphasis` contain a 'prob' key
+    which determines the probability that the corresponding transformation is applied. 
+    """
+
+    def __init__(
+        self,
+        speed: Optional[Dict[str, Any]] = None,
+        tempo: Optional[Dict[str, Any]] = None,
+        sox_effects: Optional[List[Dict[str, Any]]] = None,
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:31]: List of lists, see older unresolved comment.

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py (hunk 22 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py]: I see.

File: users/vieting/experiments/switchboard/hybrid/feat/baseline_args.py (hunk 25 above)
Comment (vieting) [reply users/vieting/experiments/switchboard/hybrid/feat/baseline_args.py:29]: Why did you mark this as resolved? It is not resolved.

**Review by vieting — APPROVED** [review]
I fixed some remaining issues myself now so that we can merge. As I said before in a comment: It would be very helpful if you look at the changes in a PR yourself with the eyes of a reviewer. Then you will see that it is strange if there are lines removed or whitespaces added for no reason. This will make the reviewers' life easier and reviewing processes faster.
